### PR TITLE
Refactor unique key check and error handling

### PR DESCRIPTION
**Pull request by dwelch0**
Changes:
* Refactor error handling within main reconcile loop to be generic and exit with proper status code during when `run-once` is true.
* Refactor uniqueKey check function

### CODE REVIEW

File: toplevel/secretsengine/secretsengine.go [hunk 1]
```diff
@@ -25,6 +27,8 @@ type entry struct {
 
 var _ vault.Item = entry{}
 
+const toplevelName = "vault_secret_engines"
```
Comment (OliviaHY) [inline toplevel/secretsengine/secretsengine.go:30]: `toplevelName` can be used on line 68 in `RegisterConfiguration`?

Comment (dwelch0) [reply toplevel/secretsengine/secretsengine.go:30]: good catch!

File: cmd/vault-manager/main.go [hunk 2]
```diff
@@ -131,6 +130,8 @@ func main() {
 		// sort configs by priority
 		sort.Sort(ByPriority(topLevelConfigs))
 
+		// used to exit with correct status from run-once execution
+		errCounter := 0
```
[the remark this answers is not in the document]
Comment (dwelch0) [reply cmd/vault-manager/main.go]: Good point. Had an idea to try and incorporate the count value but it didn't make sense. Updated to bool

**Review by OliviaHY — APPROVED** [review]

File: pkg/utils/validations.go [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package utils
+
+// generic function to validate uniqueness across slice of attribute returned by f
+func ValidKeys[T any](items []T, f func(T) string) bool {
```
Comment (OliviaHY) [inline pkg/utils/validations.go:4]: TIL

File: go.mod [hunk 4]
```diff
@@ -1,10 +1,11 @@
 module github.com/app-sre/vault-manager
 
-go 1.17
+go 1.18
```
Comment (kwilczynski) [inline go.mod:3]: Feel free to use 1.19 here already. There should be no side effects from doing so unless you have seen something?

File: pkg/utils/validations.go (hunk 3 above)
Comment (kwilczynski) [reply pkg/utils/validations.go:4]: Very nice use of modern Go! I love it!

File: tests/Dockerfile.tests [hunk 5]
```diff
@@ -1,4 +1,4 @@
-FROM registry.access.redhat.com/ubi8/go-toolset:1.17 as builder
+FROM registry.access.redhat.com/ubi8/go-toolset:1.18 as builder
```
Comment (kwilczynski) [inline tests/Dockerfile.tests:1]: Same as above. If we aren't constrained by the container image here, 1.19 could be worth considering.

File: toplevel/role/role.go [hunk 6]
```diff
@@ -114,8 +114,11 @@ func (c config) Apply(address string, entriesBytes []byte, dryRun bool, threadPo
 	}
 
 	desiredRoles := instancesToDesiredRoles[address]
-	if validateUniquenessError := validateRoleUniqueness(desiredRoles, toplevelName); validateUniquenessError != nil {
-		return validateUniquenessError
+	if unique := utils.ValidKeys(desiredRoles,
+		func(e entry) string {
+			return fmt.Sprintf("%s%s", e.Mount, e.Name)
```
[the remark this answers is not in the document]
Comment (kwilczynski) [reply toplevel/role/role.go:119]: Very nice indeed!

File: cmd/vault-manager/main.go [hunk 7]
```diff
@@ -131,6 +130,8 @@ func main() {
 		// sort configs by priority
 		sort.Sort(ByPriority(topLevelConfigs))
 
+		// used to exit with correct status from run-once execution
+		erred := false
```
Comment (kwilczynski) [inline cmd/vault-manager/main.go:134]: What if we call this "hasErrors" or "seenErrors" instead?

File: cmd/vault-manager/main.go [hunk 8]
```diff
@@ -162,7 +161,10 @@ func main() {
 		}
 
 		if runOnce {
-			return
+			if erred {
+				os.Exit(1)
```
Comment (kwilczynski) [inline cmd/vault-manager/main.go:165]: A fun trivia here. If you were to keep the counter, then you could reach out for some bit twiddling magic:

```
os.Exit((^counter + 1) >> 31 & 1)
```

This way, any non-zero value would exit with code 1, and zero would be 0.

**Review by kwilczynski — APPROVED** [review]
@dwelch0, looks good!

Very nice refactor! I love the move to leverage Go generics. 😄

**Comment by dwelch0** [comment]
@kwilczynski thank you for the great suggestion!

**Pull request merged**